nwbwidgets/placefield.py

```diff
@@ -0,0 +1,321 @@
+import numpy as np
+from scipy.ndimage import label
+from scipy.ndimage.filters import gaussian_filter, maximum_filter
+
+import matplotlib.pyplot as plt
+
+import pynwb
+from ipywidgets import widgets, BoundedFloatText, Dropdown, Checkbox, fixed
+
+from .analysis.placefields import compute_2d_firing_rate, compute_linear_firing_rate
+
+from .utils.widgets import interactive_output
+from .utils.units import get_spike_times
+from .utils.timeseries import get_timeseries_in_units, get_timeseries_tt
+from .base import vis2widget
+from .controllers import  GroupAndSortController
+
+
+## To-do
+# [X] Create PlaceFieldWidget class
+    # [X] Refactor place field calculation code to deal with nwb data type
+        # [X] Incorporate place field fxns into class
+        # [X] Change all internal attributes references
+        # [X]Change all internal method references
+
+    # [X] Get pos
+    # [X] Get time
+    # [X] Get spikes
+    # [] Get trials / epochs
+
+# [X] Submit draft PR
+
+    # [] 1D Place Field Widget
+        # [X] Incorporate nelpy package into widget
+        # [X] Speed threshold implementation
+        # [X] Add foreign group and sort controller to pick unit groups and ranges?
+        # [X] Normalized firing rate figure?
+        # [X] Add collapsed unit vizualization?
+        # [] Scale bar?
+        # [] Sort place cell tuning curves by peak firing rate position?
+        # [] Color palette control?
+
+# [] Dropdown that controls which unit
+
+# [x] Work in buttons / dropdowns / sliders to modify following parameters in place field calculation:
+# [] Different epochs
+# [x] Gaussian SD
+# [x] Speed threshold
+# [] Minimum firing rate
+# [] Place field thresh (% of local max)
+
+def route_placefield(spatial_series: pynwb.behavior.SpatialSeries):
+    print(spatial_series.data.shape)
+    if spatial_series.data.shape[1] == 2:
+        return PlaceFieldWidget(spatial_series)
+    elif spatial_series.data.shape[1] == 1:
+        return PlaceField_1D_Widget(spatial_series)
+    else:
+        print('Spatial series exceeds dimensionality for visualization')
+        return
+
+
+# Put widget rendering here
+class PlaceFieldWidget(widgets.HBox):
+
+    def __init__(self, spatial_series: pynwb.behavior.SpatialSeries, **kwargs):
+        super().__init__()
+
+
+        if hasattr(spatial_series.get_ancestor('NWBFile'),'velocity'):
+            velocity = spatial_series.get_ancestor('NWBFile').velocity
+        else:
+            velocity = []
```

Comment:
We won't be able to find velocity like this. This only really works for units and trials